File: src/agent_feed/console.py
"""Rich console output for Agent Feed."""

from __future__ import annotations

from dataclasses import asdict
import json
from pathlib import Path

from rich import box
from rich.console import Console
from rich.markdown import Markdown
from rich.panel import Panel
from rich.table import Table
from rich.text import Text

from agent_feed import __version__
from agent_feed.config import ConfigCheckReport
from agent_feed.models import CheckReport, ProjectStatus, WriteAction


console = Console()


def print_welcome() -> None:
    logo = Text()
    logo.append("AGENT\n", style="bold cyan")
    logo.append("FEED", style="bold green")

    body = Table.grid(padding=(0, 2))
    body.add_column(no_wrap=True)
    body.add_column(ratio=1)
    body.add_row(
        logo,
        "[bold white]Agent Feed[/bold white]\n"
        "[dim]A source-controlled workflow pipeline for AI coding agents.[/dim]\n\n"
        "[bold]Start[/bold]    agent-feed init\n"
        "[bold]Verify[/bold]   agent-feed check\n"
        "[bold]Inspect[/bold]  agent-feed status\n"
        "[bold]Preview[/bold]  agent-feed preview",
    )
    console.print(
        Panel.fit(
            body,
            title=f"agent-feed {__version__}",
            border_style="cyan",
            padding=(1, 2),
        )
    )


def print_error_panel(title: str, errors: list[str]) -> None:
    table = Table.grid(padding=(0, 1))
    table.add_column(width=2, no_wrap=True)
    table.add_column(ratio=1)
    for error in errors:
        table.add_row("[bold red]![/bold red]", error)
    console.print(Panel(table, title=title, border_style="red", expand=False))


def render_diff(diff: str) -> Text:
    rendered = Text()
    lines = diff.splitlines(keepends=True)
    if not lines and diff:
        lines = [diff]
    for line in lines:
        rendered.append(line, style=diff_line_style(line))
    return rendered


def diff_line_style(line: str) -> str | None:
    if line.startswith("@@"):
        return "bold cyan"
    if line.startswith("--- "):
        return "bold red"
    if line.startswith("+++ "):
        return "bold green"
    if line.startswith("-"):
        return "red"
    if line.startswith("+"):
        return "green"
    if line.startswith("diff ") or line.startswith("index "):
        return "bold"
    if line.startswith("\\"):
        return "dim"
    return None


def print_write_plan(actions: list[WriteAction], *, show_diffs: bool = False) -> None:
    print_write_plan_with_title(actions, title=write_plan_title(actions), show_diffs=show_diffs)


def print_write_plan_with_title(
    actions: list[WriteAction],
    *,
    title: str,
    show_diffs: bool = False,
) -> None:
    table = Table(
        title=title,
        box=box.SIMPLE_HEAVY,
        show_lines=False,
        header_style="bold",
    )
    table.add_column("", width=2, no_wrap=True)
    table.add_column("Action", style="bold")
    table.add_column("Path", overflow="fold")
    table.add_column("Detail", style="dim", overflow="fold")
    for action in actions:
        table.add_row(
            action_icon(action),
            styled_action(action),
            display_path(action.path),
            action.detail,
        )
    console.print(table)
    if show_diffs:
        print_diff_details(actions)


def has_diff_details(actions: list[WriteAction]) -> bool:
    return any(action.diff for action in actions)


def print_diff_details(actions: list[WriteAction]) -> None:
    for action in actions:
        if action.diff:
            console.print(
                Panel(
                    render_diff(action.diff),
                    title=f"Diff: {display_path(action.path)}",
                    border_style="yellow",
                    expand=False,
                )
            )


def print_check_report(report: CheckReport, *, as_json: bool) -> None:
    if as_json:
        payload = {
            "target": str(report.target),
            "checks": [check.value for check in report.checks],
            "ok": report.ok,
            "errors": report.errors,
            "warnings": report.warnings,
        }
        print(json.dumps(payload, indent=2, ensure_ascii=False))
        return

    if report.ok and not report.warnings:
        console.print(
            Panel.fit(
                f"[bold green]Checks passed[/bold green]\n"
                f"[dim]Target[/dim] {report.target}\n"
                f"[dim]Scope[/dim]  {', '.join(check.value for check in report.checks)}",
                title="Agent Feed",
                border_style="green",
            )
        )
    elif report.ok:
        table = Table(title="Checks Passed With Warnings", box=box.SIMPLE_HEAVY)
        table.add_column("", width=2)
        table.add_column("Type")
        table.add_column("Message", overflow="fold")
        for warning in report.warnings:
            table.add_row("?", "[yellow]warning[/yellow]", warning)
        console.print(table)
        print_next_step("Review the warnings above before the final handoff.")
    else:
        table = Table(title="Checks blocked", box=box.SIMPLE_HEAVY, header_style="bold red")
        table.add_column("", width=2)
        table.add_column("Type")
        table.add_column("Message", overflow="fold")
        for error in report.errors:
            table.add_row("!", "[red]error[/red]", error)
        for warning in report.warnings:
            table.add_row("?", "[yellow]warning[/yellow]", warning)
        console.print(table)
        print_next_step("Fix the diagnostics above, then rerun `agent-feed check`.")


def print_config_check_report(report: ConfigCheckReport, *, as_json: bool) -> None:
    if as_json:
        payload = {
            "target": str(report.target),
            "project_config": str(report.project_config),
            "user_config": str(report.user_config) if report.user_config else None,
            "ok": report.ok,
            "errors": list(report.errors),
            "warnings": list(report.warnings),
        }
        print(json.dumps(payload, indent=2, ensure_ascii=False))
        return

    if report.ok and not report.warnings:
        console.print(
            Panel.fit(
                "[bold green]Config checks passed[/bold green]\n"
                f"[dim]Project[/dim] {report.project_config}\n"
                f"[dim]User[/dim]    {report.user_config or 'not configured'}",
                title="Agent Feed",
                border_style="green",
            )
        )
        return

    table = Table(title="Config Diagnostics", box=box.SIMPLE_HEAVY, header_style="bold")
    table.add_column("", width=2)
    table.add_column("Type")
    table.add_column("Message", overflow="fold")
    for error in report.errors:
        table.add_row("!", "[red]error[/red]", error)
    for warning in report.warnings:
        table.add_row("?", "[yellow]warning[/yellow]", warning)
    console.print(table)
    if report.errors:
        print_next_step("Fix the config diagnostics above, then rerun `agent-feed config check`.")
    else:
        print_next_step("Review stale entries or rerun `agent-feed config set` to clean them.")


def print_status(status: ProjectStatus, *, as_json: bool) -> None:
    if as_json:
        payload = asdict(status)
        payload["target"] = str(status.target)
        print(json.dumps(payload, indent=2, ensure_ascii=False))
        return

    table = Table(
        title=f"Agent Feed Status: {status.target}",
        box=box.SIMPLE_HEAVY,
        header_style="bold",
    )
    table.add_column("Area")
    table.add_column("Source")
    table.add_column("State")
    table.add_row("Canonical", "AGENTS.md + .agents/", _state(status.canonical_installed))
    table.add_row(
        "Codex",
        "AGENTS.md + .agents/skills (direct)",
        _state(status.codex_ready),
    )
    table.add_row(
        "Claude",
        "CLAUDE.md + .claude/skills (generated)",
        _state(status.claude_ready),
    )
    table.add_row(
        "Cursor",
        ".cursor/rules/agent-feed.mdc (generated)",
        _state(status.cursor_ready),
    )
    console.print(table)

    if status.errors or status.warnings:
        report = Table(title="Diagnostics", box=box.SIMPLE_HEAVY)
        report.add_column("", width=2)
        report.add_column("Type")
        report.add_column("Message", overflow="fold")
        for error in status.errors:
            report.add_row("!", "[red]error[/red]", error)
        for warning in status.warnings:
            report.add_row("?", "[yellow]warning[/yellow]", warning)
        console.print(report)

    print_next_step(status_next_step(status))


def _state(ok: bool) -> str:
    return "[green]ready[/green]" if ok else "[red]blocked[/red]"


def write_plan_title(actions: list[WriteAction]) -> str:
    if not actions:
        return "No Changes"
    if any(action.severity == "warning" for action in actions):
        return "Review Required"
    if any(action.severity == "preview" for action in actions):
        return "Preview"
    return "Changes"


def action_icon(action: WriteAction) -> str:
    return {
        "warning": "!",
        "preview": "~",
        "neutral": "-",
        "success": "+",
    }[action.severity]


def styled_action(action: WriteAction) -> Text:
    styles = {
        "warning": "bold yellow",
        "preview": "cyan",
        "neutral": "dim",
        "success": "green",
    }
    return Text(action.action, style=styles[action.severity])


def display_path(path: Path) -> str:
    try:
        return str(path.relative_to(Path.cwd()))
    except ValueError:
        return str(path)


def print_next_step(message: str) -> None:
    console.print(f"[bold cyan]Next:[/bold cyan] [bold white]{message}[/bold white]")


def print_action_result(
    *,
    title: str,
    message: str,
    kind: str = "info",
    detail: str | None = None,
) -> None:
    styles = {
        "success": ("green", "bold green"),
        "warning": ("yellow", "bold yellow"),
        "error": ("red", "bold red"),
        "info": ("cyan", "bold cyan"),
    }
    border_style, heading_style = styles.get(kind, styles["info"])
    body = f"[{heading_style}]{message}[/{heading_style}]"
    if detail:
        body += f"\n[dim]{detail}[/dim]"
    console.print(Panel.fit(body, title=title, border_style=border_style))


def print_recommended_command(message: str, command: str, *, path: str | None = None) -> None:
    text = Text()
    text.append(message, style="bold cyan")
    if path:
        text.append(" Edit ")
        text.append(path, style="blue italic")
        text.append(", then run ")
    else:
        text.append(" Run ")
    text.append(command, style="bold green")
    text.append(".")
    console.print(text)


def print_markdown_panel(title: str, body: str, *, border_style: str = "blue") -> None:
    console.print(Panel(Markdown(body), title=title, border_style=border_style, expand=False))


def print_stale_project_cleanup(config_file: Path, project_roots: tuple[Path, ...]) -> None:
    metadata = Table(
        box=None,
        show_header=False,
        expand=True,
        padding=(0, 1),
    )
    metadata.add_column("Field", style="dim", no_wrap=True)
    metadata.add_column("Value", overflow="fold", ratio=1)
    metadata.add_row("Config", str(config_file))

    table = Table(
        box=box.SIMPLE,
        show_header=True,
        header_style="bold",
        expand=True,
    )
    table.add_column("#", justify="right", style="dim", width=3)
    table.add_column("Missing project root", overflow="fold", ratio=1)
    for index, project_root in enumerate(project_roots, start=1):
        table.add_row(str(index), str(project_root))

    body = Table.grid(expand=True, padding=(0, 1))
    body.add_column(ratio=1)
    body.add_row(
        "Agent Feed found project records in the user-level config whose directories no "
        "longer exist. Removing them only cleans stale trust metadata; it does not touch "
        "project files."
    )
    body.add_row(metadata)
    body.add_row(table)
    console.print(
        Panel(
            body,
            title="Stale Project Entries",
            border_style="yellow",
            expand=False,
        )
    )


def status_next_step(status: ProjectStatus) -> str:
    if not status.canonical_installed:
        return "Run `agent-feed init` in this project."
    if status.errors:
        return "Run `agent-feed check --checks all` for the full failure list."
    if status.warnings:
        return "Review the warnings above, then run `agent-feed preview` before updating."
    return "Run `agent-feed preview` to inspect managed drift before updating."


def print_diff_hint(
    *,
    command: str,
    interactive: bool,
) -> None:
    if interactive:
        message = (
            f"[bold cyan]Diff details:[/bold cyan] press [bold green]v[/bold green] "
            f"to show all diffs, or press any other key to exit. "
            f"Script mode: [bold green]{command}[/bold green]."
        )
    else:
        message = (
            f"[bold cyan]Diff details:[/bold cyan] rerun [bold green]{command}[/bold green]."
        )
    console.print(message)

File: src/node/src/console.ts
import { stderr, stdout } from "node:process";

import chalk from "chalk";
import Table from "cli-table3";

import type { CheckReport, ProjectStatus } from "./checks.js";
import type { ConfigCheckReport } from "./config.js";
import type { WriteAction } from "./template.js";
import { VERSION } from "./version.js";

type Stream = "stdout" | "stderr";
type Colorizer = (value: string) => string;

export function printWelcome(): void {
  const logo = `${chalk.bold.cyan("AGENT")}\n${chalk.bold.green("FEED")}`;
  const body = [
    `${chalk.bold.white("Agent Feed")}`,
    chalk.dim("A source-controlled workflow pipeline for AI coding agents."),
    "",
    `${chalk.bold("Start")}    ${command("agent-feed init")}`,
    `${chalk.bold("Verify")}   ${command("agent-feed check")}`,
    `${chalk.bold("Inspect")}  ${command("agent-feed status")}`,
    `${chalk.bold("Preview")}  ${command("agent-feed preview")}`,
  ].join("\n");
  printBox({
    title: `agent-feed ${VERSION}`,
    body: sideBySide(logo, body, 4),
    color: chalk.cyan,
  });
}

export function printErrorPanel(title: string, errors: string[], stream: Stream = "stderr"): void {
  const rows = errors.map((error) => `${chalk.bold.red("!")} ${styleMessage(error)}`);
  printBox({ title, body: rows.join("\n"), color: chalk.red, stream });
}

export function printWritePlan(actions: WriteAction[], options: { showDiffs?: boolean; title?: string } = {}): void {
  const table = new Table({
    head: ["", "Action", "Path", "Detail"].map((item) => chalk.bold(item)),
    style: { head: [], border: ["gray"] },
    wordWrap: false,
  });
  for (const action of actions) {
    table.push([
      actionIcon(action),
      styledAction(action),
      displayPath(action.path),
      chalk.dim(action.detail ?? ""),
    ]);
  }
  const title = options.title ?? writePlanTitle(actions);
  stdout.write(`${chalk.bold(title)}\n${table.toString()}\n`);
  if (options.showDiffs) {
    printDiffDetails(actions);
  }
}

export function hasDiffDetails(actions: WriteAction[]): boolean {
  return actions.some((action) => Boolean(action.diff));
}

export function printDiffDetails(actions: WriteAction[]): void {
  for (const action of actions) {
    if (!action.diff) {
      continue;
    }
    printBox({
      title: `Diff: ${displayPath(action.path)}`,
      body: renderDiff(action.diff),
      color: chalk.yellow,
    });
  }
}

export function renderDiff(diff: string): string {
  return diff
    .split(/\r?\n/)
    .map((line) => diffLineStyle(line)(line))
    .join("\n");
}

export function printCheckReport(report: CheckReport, options: { asJson: boolean }): void {
  if (options.asJson) {
    stdout.write(`${JSON.stringify(report, null, 2)}\n`);
    return;
  }
  if (report.ok && report.warnings.length === 0) {
    printBox({
      title: "Agent Feed",
      body: [
        chalk.bold.green("Checks passed"),
        `${chalk.dim("Target")} ${displayPath(report.target)}`,
        `${chalk.dim("Scope")}  ${report.checks.join(", ")}`,
      ].join("\n"),
      color: chalk.green,
    });
    return;
  }
  const rows = [
    ...report.errors.map((message) => ["!", chalk.red("error"), styleMessage(message)]),
    ...report.warnings.map((message) => ["?", chalk.yellow("warning"), styleMessage(message)]),
  ];
  printDiagnostics(report.ok ? "Checks Passed With Warnings" : "Checks blocked", rows, report.ok ? "stdout" : "stderr");
  printNextStep(checkNextStep(report.errors, report.warnings), report.ok ? "stdout" : "stderr");
}

export function printConfigCheckReport(report: ConfigCheckReport): void {
  if (report.errors.length === 0 && report.warnings.length === 0) {
    printBox({
      title: "Agent Feed",
      body: chalk.bold.green("Config checks passed"),
      color: chalk.green,
    });
    return;
  }
  const rows = [
    ...report.errors.map((message) => ["!", chalk.red("error"), styleMessage(message)]),
    ...report.warnings.map((message) => ["?", chalk.yellow("warning"), styleMessage(message)]),
  ];
  printDiagnostics("Config Diagnostics", rows, report.errors.length > 0 ? "stderr" : "stderr");
  printNextStep(configNextStep(report.errors, report.warnings), "stderr");
}

export function printStatus(status: ProjectStatus, options: { asJson: boolean }): void {
  if (options.asJson) {
    stdout.write(`${JSON.stringify(status, null, 2)}\n`);
    return;
  }
  const table = new Table({
    head: ["Area", "Source", "State"].map((item) => chalk.bold(item)),
    style: { head: [], border: ["gray"] },
    wordWrap: false,
  });
  table.push(
    ["Canonical", "AGENTS.md + .agents/", state(status.canonical_installed)],
    ["Codex", "AGENTS.md + .agents/skills (direct)", state(status.codex_ready)],
    ["Claude", "CLAUDE.md + .claude/skills (generated)", state(status.claude_ready)],
    ["Cursor", ".cursor/rules/agent-feed.mdc (generated)", state(status.cursor_ready)],
  );
  stdout.write(`${chalk.bold(`Agent Feed Status: ${displayPath(status.target)}`)}\n${table.toString()}\n`);
  if (status.errors.length > 0 || status.warnings.length > 0) {
    printDiagnostics(
      "Diagnostics",
      [
        ...status.errors.map((message) => ["!", chalk.red("error"), styleMessage(message)]),
        ...status.warnings.map((message) => ["?", chalk.yellow("warning"), styleMessage(message)]),
      ],
      status.errors.length > 0 ? "stderr" : "stdout",
    );
  }
  printNextStep(statusNextStep(status));
}

export function printInspectionPlan(target: string, actions: WriteAction[], interactive: boolean): void {
  printWritePlan(actions, { title: `Agent Feed Inspection: ${displayPath(target)}` });
  if (!hasDiffDetails(actions)) {
    return;
  }
  printDiffHint({
    command: `agent-feed preview ${target}`,
    interactive,
  });
}

export function printDiffHint(options: { command: string; interactive: boolean }): void {
  if (options.interactive) {
    stdout.write(
      `${chalk.bold.cyan("Diff details:")} press ${chalk.bold.green("v")} to show all diffs, or press any other key to exit. Script mode: ${command(options.command)}.\n`,
    );
    return;
  }
  stdout.write(`${chalk.bold.cyan("Diff details:")} rerun ${command(options.command)}.\n`);
}

export function printActionResult(options: {
  title: string;
  message: string;
  kind?: "success" | "warning" | "error" | "info";
  detail?: string;
}): void {
  const color =
    options.kind === "success"
      ? chalk.green
      : options.kind === "warning"
        ? chalk.yellow
        : options.kind === "error"
          ? chalk.red
          : chalk.cyan;
  const stream = options.kind === "error" ? "stderr" : "stdout";
  const body = [chalk.bold(color(options.message)), options.detail ? chalk.dim(styleMessage(options.detail)) : ""]
    .filter(Boolean)
    .join("\n");
  printBox({ title: options.title, body, color, stream });
}

export function printPanel(
  title: string,
  body: string,
  options: { kind?: "success" | "warning" | "error" | "info"; stream?: Stream } = {},
): void {
  const color =
    options.kind === "success"
      ? chalk.green
      : options.kind === "warning"
        ? chalk.yellow
        : options.kind === "error"
          ? chalk.red
          : chalk.cyan;
  printBox({ title, body, color, stream: options.stream });
}

export function printRecommendedCommand(message: string, value: string, options: { path?: string } = {}): void {
  const pieces = [chalk.bold.cyan(message)];
  if (options.path) {
    pieces.push(" Edit ", pathStyle(options.path), ", then run ");
  } else {
    pieces.push(" Run ");
  }
  pieces.push(command(value), ".");
  stdout.write(`${pieces.join("")}\n`);
}

export function printWarning(message: string, stream: Stream = "stderr"): void {
  write(stream, `${chalk.bold.yellow("Warning:")} ${styleMessage(message)}\n`);
}

export function printInfo(message: string): void {
  stdout.write(`${styleMessage(message)}\n`);
}

export function printNextStep(message: string, stream: Stream = "stdout"): void {
  write(stream, `${chalk.bold.cyan("Next:")} ${styleMessage(message)}\n`);
}

export function styleMessage(message: string): string {
  let styled = message.replace(/`([^`]+)`/g, (_match, value: string) => command(value));
  styled = styled.replace(/((?:\.{1,2}|~|\/|[A-Za-z]:\\)[^\s,;:]+|[A-Za-z0-9_.-]+\/[^\s,;:]+)/g, (match) => {
    if (match.includes("\u001B[")) {
      return match;
    }
    return pathStyle(match);
  });
  return styled;
}

export function command(value: string): string {
  return chalk.bold.green(value);
}

export function pathStyle(value: string): string {
  return chalk.italic.blue(value);
}

function printDiagnostics(title: string, rows: string[][], stream: Stream): void {
  const table = new Table({
    head: ["", "Type", "Message"].map((item) => chalk.bold(item)),
    style: { head: [], border: ["gray"] },
    wordWrap: false,
  });
  for (const row of rows) {
    table.push(row);
  }
  write(stream, `${chalk.bold(title)}\n${table.toString()}\n`);
}

function writePlanTitle(actions: WriteAction[]): string {
  if (actions.length === 0) {
    return "No Changes";
  }
  if (actions.some((action) => actionSeverity(action) === "warning")) {
    return "Review Required";
  }
  if (actions.some((action) => actionSeverity(action) === "preview")) {
    return "Preview";
  }
  return "Changes";
}

function actionIcon(action: WriteAction): string {
  const severity = actionSeverity(action);
  if (severity === "warning") {
    return chalk.bold.yellow("!");
  }
  if (severity === "preview") {
    return chalk.cyan("~");
  }
  if (severity === "success") {
    return chalk.green("+");
  }
  return chalk.dim("-");
}

function styledAction(action: WriteAction): string {
  const severity = actionSeverity(action);
  if (severity === "warning") {
    return chalk.bold.yellow(action.action);
  }
  if (severity === "preview") {
    return chalk.cyan(action.action);
  }
  if (severity === "success") {
    return chalk.green(action.action);
  }
  return chalk.dim(action.action);
}

function actionSeverity(action: WriteAction): "warning" | "preview" | "neutral" | "success" {
  const value = action.action.toLowerCase();
  if (value.includes("blocked") || value.includes("warn") || value.includes("error")) {
    return "warning";
  }
  if (value.startsWith("would")) {
    return "preview";
  }
  if (["create", "update", "sync", "delete", "backup", "ok"].includes(value)) {
    return "success";
  }
  return "neutral";
}

function displayPath(value: string): string {
  const cwd = process.cwd();
  if (value === cwd) {
    return ".";
  }
  if (value.startsWith(`${cwd}/`)) {
    return value.slice(cwd.length + 1);
  }
  return value;
}

function diffLineStyle(line: string): (value: string) => string {
  if (line.startsWith("@@")) {
    return chalk.bold.cyan;
  }
  if (line.startsWith("--- ")) {
    return chalk.bold.red;
  }
  if (line.startsWith("+++ ")) {
    return chalk.bold.green;
  }
  if (line.startsWith("-")) {
    return chalk.red;
  }
  if (line.startsWith("+")) {
    return chalk.green;
  }
  if (line.startsWith("diff ") || line.startsWith("index ")) {
    return chalk.bold;
  }
  if (line.startsWith("\\")) {
    return chalk.dim;
  }
  return (value: string) => value;
}

function state(ok: boolean): string {
  return ok ? chalk.green("ready") : chalk.red("blocked");
}

function checkNextStep(errors: string[], warnings: string[]): string {
  const diagnostics = [...errors, ...warnings].join("\n");
  if (diagnostics.includes("stale project entry")) {
    return "Run `agent-feed config prune`, then rerun `agent-feed check -a`.";
  }
  if (diagnostics.includes("AGENT_FEED_HOME") || diagnostics.includes("user-level Agent Feed config")) {
    return "Run `agent-feed env setup`, then rerun `agent-feed check -a`.";
  }
  if (errors.length > 0) {
    return "Fix the diagnostics above, then rerun `agent-feed check -a`.";
  }
  return "Review the warnings above before the final handoff.";
}

function configNextStep(errors: string[], warnings: string[]): string {
  const diagnostics = [...errors, ...warnings].join("\n");
  if (diagnostics.includes("stale project entry")) {
    return "Run `agent-feed config prune`, then rerun `agent-feed config check`.";
  }
  if (diagnostics.includes("AGENT_FEED_HOME") || diagnostics.includes("user-level Agent Feed config")) {
    return "Run `agent-feed env setup`, then rerun `agent-feed config check`.";
  }
  if (errors.length > 0) {
    return "Fix the config diagnostics above, then rerun `agent-feed config check`.";
  }
  return "Review the config warnings above.";
}

function statusNextStep(status: ProjectStatus): string {
  if (!status.canonical_installed) {
    return "Run `agent-feed init` in this project.";
  }
  if (status.errors.length > 0) {
    return "Run `agent-feed check -a` for the full failure list.";
  }
  if (status.warnings.length > 0) {
    return "Review the warnings above, then run `agent-feed preview` before updating.";
  }
  return "Run `agent-feed preview` to inspect managed drift before updating.";
}

function printBox(options: {
  title: string;
  body: string;
  color: Colorizer;
  stream?: Stream;
}): void {
  const lines = options.body.split("\n");
  const width = Math.max(
    stripAnsi(options.title).length + 4,
    ...lines.map((line) => stripAnsi(line).length + 4),
  );
  const top = `${options.color("╭")}${options.color("─".repeat(width - 2))}${options.color("╮")}`;
  const title = ` ${options.title} `;
  const titleLine = `${options.color("│")}${center(title, width - 2)}${options.color("│")}`;
  const bottom = `${options.color("╰")}${options.color("─".repeat(width - 2))}${options.color("╯")}`;
  const body = lines
    .map((line) => `${options.color("│")} ${line}${" ".repeat(width - stripAnsi(line).length - 3)}${options.color("│")}`)
    .join("\n");
  write(options.stream ?? "stdout", `${top}\n${titleLine}\n${body}\n${bottom}\n`);
}

function sideBySide(left: string, right: string, gap: number): string {
  const leftLines = left.split("\n");
  const rightLines = right.split("\n");
  const leftWidth = Math.max(...leftLines.map((line) => stripAnsi(line).length));
  const height = Math.max(leftLines.length, rightLines.length);
  const rows: string[] = [];
  for (let index = 0; index < height; index += 1) {
    const leftLine = leftLines[index] ?? "";
    const rightLine = rightLines[index] ?? "";
    rows.push(`${leftLine}${" ".repeat(leftWidth - stripAnsi(leftLine).length + gap)}${rightLine}`);
  }
  return rows.join("\n");
}

function center(value: string, width: number): string {
  const visible = stripAnsi(value).length;
  const padding = Math.max(width - visible, 0);
  const left = Math.floor(padding / 2);
  const right = padding - left;
  return `${" ".repeat(left)}${value}${" ".repeat(right)}`;
}

function write(stream: Stream, value: string): void {
  (stream === "stderr" ? stderr : stdout).write(value);
}

function stripAnsi(value: string): string {
  return value.replace(/\u001B\[[0-9;]*m/g, "");
}
